airflow/providers/amazon/aws/operators/athena.py

```diff
@@ -163,3 +167,132 @@ def on_kill(self) -> None:
+        parse_result = sql_parser.parse(sql=self.query)
+
+        if not parse_result:
+            return OperatorLineage(job_facets=job_facets)
+
+        run_facets: dict[str, BaseFacet] = {}
+        if parse_result.errors:
+            run_facets["extractionError"] = ExtractionErrorRunFacet(
+                totalTasks=len(self.query) if isinstance(self.query, list) else 1,
+                failedTasks=len(parse_result.errors),
+                errors=[
+                    ExtractionError(
+                        errorMessage=error.message,
+                        stackTrace=None,
+                        task=error.origin_statement,
+                        taskNumber=error.index,
+                    )
+                    for error in parse_result.errors
+                ],
+            )
+
+        inputs: list[Dataset] = list(
+            filter(
+                None,
+                [
+                    self.get_openlineage_dataset(table.schema or self.database, table.name)
+                    for table in parse_result.in_tables
+                ],
+            )
+        )
+
+        # Athena can output query result to a new table with CTAS query.
+        # cf. https://docs.aws.amazon.com/athena/latest/ug/ctas.html
+        outputs: list[Dataset] = list(
+            filter(
+                None,
+                [
+                    self.get_openlineage_dataset(table.schema or self.database, table.name)
+                    for table in parse_result.out_tables
+                ],
+            )
+        )
+
+        # In addition to CTAS query, it's also possible to specify output location on S3
+        # with a mandatory parameter, OutputLocation in ResultConfiguration.
+        # cf. https://docs.aws.amazon.com/athena/latest/APIReference/API_ResultConfiguration.html#athena-Type-ResultConfiguration-OutputLocation  # noqa: E501
+        #
+        # Depending on the query type and the external_location property in the CTAS query,
+        # its behavior changes as follows:
+        #
+        # * Normal SELECT statement
+        #   -> The result is put into output_location as files rather than a table.
+        #
+        # * CTAS statement without external_location (`CREATE TABLE ... AS SELECT ...`)
+        #   -> The result is put into output_location as a table,
+        #      that is, both metadata files and data files are in there.
+        #
+        # * CTAS statement with external_location
+        #   (`CREATE TABLE ... WITH (external_location='s3://bucket/key') AS SELECT ...`)
+        #   -> The result is output as a table, but metadata and data files are
+        #      separated into output_location and external_location respectively.
+        #
+        # For the last case, output_location may be unnecessary as OL's output information,
+        # but we keep it as of now since it may be useful for some purpose.
```

Comment:
> Also I don't think you can run Athena query without specifying output_location?

Technically you could run Athena query without specifying `OutputLocation` in `ResultConfiguration` by setting up output location in `Workgroup`. However in AthenaOperator `output_location` it is mandatory field even though it is not mandatory in boto3 / AWS API

Reply:
So we should change it.
If user doesnt set it right I guess boto3/athena api will raise exception

Reply:
> If user doesnt set it right I guess boto3/athena api will raise exception

Yeas that is true. For example two cases

With this settings on `primary` group, which is used by default in boto3 / AthenaOperator
![image](https://github.com/apache/airflow/assets/3998685/98a5c2bf-424c-4b5c-b233-3fb1c8ea90cf)

This simple code would run without any issues thought `boto3`

```python
import boto3

session = boto3.session.Session(...)
client = session.client(service_name="athena")

response = client.start_query_execution(QueryString="SELECT 1")
assert response
```

However if WorkGroup doesn't assign any output location, like this one:
![image](https://github.com/apache/airflow/assets/3998685/8eaedf77-e72d-4b20-a03c-a0cc0d9f0a90)

Then it would raise an error
```python
import boto3

session = boto3.session.Session(...)
client = session.client(service_name="athena", WorkGroup="test_workgroup")

response = client.start_query_execution(QueryString="SELECT 1")
```

```console
botocore.errorfactory.InvalidRequestException: An error occurred (InvalidRequestException) when calling the StartQueryExecution operation: No output location provided. An output location is required either through the Workgroup result configuration setting or as an API input.
```

Reply:
> 2. This information seems to be user facing not developer facing thus it should be in the docs not as comment in the code.

I agree. I'll remove those comments and move them to docs.

Regarding output location - whatever the change would be (seems like `output_location` should not be required) - is there a way to retrieve information from WorkGroup? Am I correct this is the right information: https://docs.aws.amazon.com/athena/latest/APIReference/API_ResultConfiguration.html#athena-Type-ResultConfiguration-OutputLocation?

Reply:
I guess you should better grab this information from [GetQueryExecution](https://docs.aws.amazon.com/athena/latest/APIReference/API_GetQueryExecution.html) API ([`Athena.Client.get_query_execution`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/athena/client/get_query_execution.html))

That is quite difficult (I think even impossible) to resolve it without query to this API endpoint, workgroup might have settings to always overwrite API settings for output locations

```python
import boto3

session = boto3.session.Session(...)
client = session.client(service_name="athena")

query_execution_id = client.start_query_execution(QueryString="SELECT 1")["QueryExecutionId"]
response = client.get_query_execution(QueryExecutionId=query_execution_id)
result_configuration = response["QueryExecution"]["ResultConfiguration"]
```
```json
{
  "OutputLocation": "s3://foo-bar/70c5fe8e-eb6c-4d98-b584-23b0120d4671.csv",
  "EncryptionConfiguration": {
    "EncryptionOption": "SSE_S3"
  }
}
```

Reply:
I made some more research. It seems like there might be different cases, e.g. set `external_location` that overrides setting from workgroup (or raise error if enforcing query location is enabled).
This all seems a bit too complicated but on the other hand [get_query_execution](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/athena/client/get_query_execution.html) returns `OutputLocation` as well. Therefore I think it will be best to save the property when checking query status. WDYT?

Reply:
@eladkal  @Taragolis I've let myself introduce caching to `get_query_execution` caching with backwards compatibility - user needs to explicitly set `use_cached_response` argument.

`output_location` is now set when query succeeds and reused later in OL.